File: Hodor/hodor-main/sdk/nodejs/example-voice.js
/**
 * Voice-activated unlock example using Vosk (offline speech recognition).
 *
 * Requirements:
 *   npm install vosk mic
 *   Download a Vosk model from https://alphacephei.com/vosk/models
 *   and extract to ./vosk-model-small-en-us (or adjust MODEL_PATH).
 *
 * Usage:
 *   node example-voice.js <username> <password> [domain]
 */

"use strict";

const vosk = require("vosk");
const mic = require("mic");
const { CredentialProviderClient } = require("./credential-provider");

const MODEL_PATH = "./vosk-model-small-en-us";
const TRIGGER_PHRASES = ["unlock", "unlock computer", "open sesame"];

async function voiceUnlock(username, password, domain = ".") {
  const client = new CredentialProviderClient();

  if (!require("fs").existsSync(MODEL_PATH)) {
    console.error(`Model not found at ${MODEL_PATH}.`);
    console.error("Download from https://alphacephei.com/vosk/models");
    process.exit(1);
  }

  vosk.setLogLevel(-1);
  const model = new vosk.Model(MODEL_PATH);
  const recognizer = new vosk.Recognizer({ model, sampleRate: 16000 });

  const micInstance = mic({
    rate: "16000",
    channels: "1",
    encoding: "signed-integer",
    bitwidth: "16",
  });

  const micStream = micInstance.getAudioStream();
  console.log('Listening. Say "unlock computer" to unlock the workstation.');
  console.log("Press Ctrl+C to stop.\n");
  micInstance.start();

  micStream.on("data", async (data) => {
    if (recognizer.acceptWaveform(data)) {
      const { text } = recognizer.result();
      if (!text) return;

      console.log(`Heard: "${text}"`);

      const matched = TRIGGER_PHRASES.some((phrase) =>
        text.toLowerCase().includes(phrase)
      );
      if (matched) {
        micInstance.stop();
        console.log("Voice command matched. Sending unlock...");
        try {
          const result = await client.unlockWithRetry(
            username,
            password,
            domain
          );
          console.log(`Result: ${result.response}`);
        } catch (err) {
          console.error(`Unlock failed: ${err.message}`);
        }
        process.exit(0);
      }
    }
  });
}

const args = process.argv.slice(2);
if (args.length < 2) {
  console.log("Usage: node example-voice.js <username> <password> [domain]");
  process.exit(1);
}
voiceUnlock(args[0], args[1], args[2] || ".");

File: Hodor/hodor-main/sdk/python/example_voice.py
"""
Voice-activated unlock example.

Listens for the phrase "unlock computer" using the SpeechRecognition
library, then sends an UNLOCK command to the credential provider.

Requirements:
    pip install SpeechRecognition pyaudio

Usage:
    python example_voice.py <username> <password> [domain]
"""

import sys
import speech_recognition as sr
from credential_provider import CredentialProviderClient


def voice_unlock(username: str, password: str, domain: str = ".") -> None:
    client = CredentialProviderClient()
    recognizer = sr.Recognizer()
    mic = sr.Microphone()

    print("Adjusting for ambient noise...")
    with mic as source:
        recognizer.adjust_for_ambient_noise(source, duration=1)

    print('Listening. Say "unlock computer" to unlock the workstation.')

    while True:
        with mic as source:
            audio = recognizer.listen(source, phrase_time_limit=5)

        try:
            text = recognizer.recognize_google(audio).lower()
            print(f"Heard: {text}")
        except sr.UnknownValueError:
            continue
        except sr.RequestError as exc:
            print(f"Speech API error: {exc}")
            continue

        if "unlock" in text:
            print("Voice command matched. Sending unlock...")
            result = client.unlock_with_retry(username, password, domain)
            print(f"Result: {result.response}")
            if result.success:
                break
            print("Unlock was not successful. Listening again...")


if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python example_voice.py <username> <password> [domain]")
        sys.exit(1)
    voice_unlock(
        sys.argv[1],
        sys.argv[2],
        sys.argv[3] if len(sys.argv) > 3 else ".",
    )
